### PR TITLE
Add sink.uri

### LINKED ISSUE #198: Replace the sink.ref in events api with function ingress
**Description**
Since we already have the Function Ingress, we can use it instead of `sink.ref` in the events API.

**Additional context**
If not, we would have to configure the Knative Serving network layer (e.g. kourier)
Using the Function Ingress can help us shield the underlying complex implementation for the user.

### PR DESCRIPTION
**Pull request by arugal**
The solutions are as follows:

```go
type SinkSpec struct {
 	Ref *Reference `json:"ref,omitempty"`
 	Uri *string    `json:"uri,omitempty"`
}
 
type Reference struct {
 	Kind       string `json:"kind"`
 	Namespace  string `json:"namespace,omitempty"`
 	Name       string `json:"name"`
 	APIVersion string `json:"apiVersion"`
}
```
 
`Reference` provides references to Knative Service and Function resources from which to retrieve function URLs `Uri` is used to pass the URL directly
 
When using the `Sink`, you must use one of these attributes, and in priority `Uri` > `Reference`

Resolves #198 

Signed-off-by: zhangwei <zhangwei24@apache.org>

### CODE REVIEW

File: apis/events/v1alpha1/eventsource_types.go [hunk 1]
```diff
@@ -52,9 +52,10 @@ type EventSourceSpec struct {
 	Build *openfunction.BuildImpl `json:"build,omitempty"`
 }
 
-// SinkSpec describes an event source for the Kafka.
+// SinkSpec describes an event source for the Kafka, priority Uri > Reference.
```
Comment (benjaminhuo) [inline apis/events/v1alpha1/eventsource_types.go:55]: @arugal Thanks for the PR, the test part is professional.

The original comment is wrong actually, would you help to correct it?
`// SinkSpec describes an event source for the Kafka, priority Uri > Reference.`
=>
`// SinkSpec specifies the receiver of the events an EventSource received, sinks in the Uri format have higher priority than sinks in Reference format.`

**Comment by benjaminhuo** [comment]
This PR looks good to me now.
After we add an URL to the function status as described in https://github.com/OpenFunction/OpenFunction/issues/210, we can update the reference part to support the OpenFunction reference to retrieve the function URL direct.

File: apis/events/v1alpha1/eventsource_types.go (hunk 1 above)
Comment (arugal) [reply apis/events/v1alpha1/eventsource_types.go]: Done

File: controllers/events/event.go [hunk 2]
```diff
@@ -172,25 +172,38 @@ func newSinkComponentSpec(c client.Client, log logr.Logger, ref *ofevent.Referen
...
+		namespace = resource.GetNamespace()
+	} else {
+		var ksvc kservingv1.Service
+		if err := c.Get(ctx, types.NamespacedName{Namespace: sink.Ref.Namespace, Name: sink.Ref.Name}, &ksvc); err != nil {
+			log.Error(err, "Failed to find Knative Service", "namespace", sink.Ref.Namespace, "name", sink.Ref.Name)
+			return nil, err
+		}
+		url = ksvc.Status.URL.String()
+		namespace = sink.Ref.Namespace
+	}
```
Comment (tpiperatgod) [inline controllers/events/event.go:192]: I think we also need to add support for Function in this section, what do you think?

Comment (arugal) [reply controllers/events/event.go:192]:
```yaml
apiVersion: core.openfunction.io/v1alpha2
kind: Function
name: function-sample
namespace: default
```

like this ?

**Review by benjaminhuo — APPROVED** [review]

**Comment by benjaminhuo** [comment]
> This PR looks good to me now. After we add an URL to the function status as described in #210, we can update the reference part to support using the OpenFunction reference to retrieve the function URL direct.

@tpiperatgod we can add URL to function status and then update this part

File: controllers/events/event.go (hunk 2 above)
Comment (tpiperatgod) [reply controllers/events/event.go:192]: yes

**Comment by benjaminhuo** [comment]
We already have an URL field in the function status in v0.5.0, so the reference can be an OpenFunction's own `Function` CRD now.

https://github.com/OpenFunction/OpenFunction/blob/main/apis/core/v1alpha2/function_types.go#L202

**Comment by arugal** [comment]
> We already have an URL field in the function status in v0.5.0, so the reference can be an OpenFunction's own `Function` CRD now. So it's ok to add Function reference now @arugal , my appology https://github.com/OpenFunction/OpenFunction/blob/main/apis/core/v1alpha2/function_types.go#L202

I will submit a new PR

**Pull request merged**